### PR TITLE
ENT-1384/ENT-1461 Add Enterprise Coupon Code Revoke API

### PR DESCRIPTION
**Pull request by brittneyexline**
**Description**
This PR introduces an api endpoint for revoking codes from users. The endpoint accepts an email template, and a list of email/code combinations. Each of these is evaluated independently, so if one email/code combination is not valid, others specified will still go through if they are valid. The response will indicate the success of each email/code pair sent. Upon success, all active assignments for that code/user are set to revoked status.

**Testing Instructions**
Deployed on the business sandbox.

Hitting the endpoint with a code that is not part of the Coupon should fail.
Request:
```POST /api/v2/enterprise/coupons/419/revoke/ HTTP/1.1
Host: ecommerce-business.sandbox.edx.org
Authorization: JWT {jwt_token}
Content-Type: application/json
{"template":"", "assignments":[{"email":"bexline+revoke@edx.org", "code":"abcdef"}]}
```

Response:
```[
    {
        "non_field_errors": [
            "Code abcdef is not associated with this Coupon"
        ]
    }
]
```

Hitting the endpoint with a code/email that do not have an active assignment should fail.
Request:
```POST /api/v2/enterprise/coupons/419/revoke/ HTTP/1.1
Host: ecommerce-business.sandbox.edx.org
Authorization: JWT {jwt_token}
Content-Type: application/json
{"template":"", "assignments":[{"email":"bexline+revoke@edx.org", "code":"TDXSDBTU3YSVCK3X"}]}
```

Response:
```[
    {
        "non_field_errors": [
            "No assignments exist for user bexline+revoke@edx.org and code TDXSDBTU3YSVCK3X"
        ]
    }
]
```

Hitting the endpoint for SINGLE_USE or ONCE_PER_CUSTOMER Coupons should set exactly one OfferAssignment object to have the status of REVOKED. This is because only one active OfferAssignment should exist for that email/code combination.
Assign a code to a user:
```POST /api/v2/enterprise/coupons/419/assign/ HTTP/1.1
Host: ecommerce-business.sandbox.edx.org
Authorization: JWT {JWT_TOKEN}
Content-Type: application/json
cache-control: no-cache
{"template":"", "emails": ["bexline@edx.org”]}
{
    "offer_assignments": [
        {
            "status": "email_dispatch_failed",
            "code": "WYZIOEZT2N7VYJ5K",
            "id": 47,
            "user_email": "bexline@edx.org"
        }
    ]
}
```

Revoke the code:
``` POST /api/v2/enterprise/coupons/419/revoke/ HTTP/1.1
Host: ecommerce-business.sandbox.edx.org
Authorization: JWT {JWT_TOKEN}
Content-Type: application/json
{"template":"", "assignments": [{"email":"bexline@edx.org", "code": "WYZIOEZT2N7VYJ5K"}]}
[
    {
        "code": "WYZIOEZT2N7VYJ5K",
        "email": "bexline@edx.org",
        "detail": "success"
    }
]
```

Once this code/email have been revoked, the code should be assignable and redeemable by a different user.

Hitting the endpoint for MULTI_USE or MULTI_USE_PER_CUSTOMER Coupons should set all active OfferAssignment objects for that code/email combination to have the status of REVOKED. 
Assign a MULTI_USE_PER_CUSTOMER code:
```POST /api/v2/enterprise/coupons/399/assign/ HTTP/1.1
Host: ecommerce-business.sandbox.edx.org
Authorization: JWT {jwt_token}
Content-Type: application/json
{"template":"", "emails":["bexline+revoke@edx.org”]}
{
    "offer_assignments": [
        {
            "status": "email_dispatch_failed",
            "code": "32XIB7PHKBHR5S7L",
            "id": 49,
            "user_email": "bexline+revoke@edx.org"
        },
        {
            "status": "email_dispatch_failed",
            "code": "32XIB7PHKBHR5S7L",
            "id": 50,
            "user_email": "bexline+revoke@edx.org"
        },
        {
            "status": "email_dispatch_failed",
            "code": "32XIB7PHKBHR5S7L",
            "id": 51,
            "user_email": "bexline+revoke@edx.org"
        },
        {
            "status": "email_dispatch_failed",
            "code": "32XIB7PHKBHR5S7L",
            "id": 52,
            "user_email": "bexline+revoke@edx.org"
        },
        {
            "status": "email_dispatch_failed",
            "code": "32XIB7PHKBHR5S7L",
            "id": 53,
            "user_email": "bexline+revoke@edx.org"
        }
    ]
}
```
Revoke the code:
```POST /api/v2/enterprise/coupons/399/revoke/ HTTP/1.1
Host: ecommerce-business.sandbox.edx.org
Authorization: JWT {jwt_token}
Content-Type: application/json
{"template":"", "assignments":[{"email":"bexline+revoke@edx.org", "code":"32XIB7PHKBHR5S7L"}]}
[
    {
        "code": "32XIB7PHKBHR5S7L",
        "email": "bexline+revoke@edx.org",
        "detail": "success"
    }
]
```

Once this code/email have been revoked, MULTI_USE codes should be assignable and redeemable, MUTLI_USE_PER_CUSTOMER codes should not be assignable or redeemable.

### CODE REVIEW

File: ecommerce/extensions/offer/utils.py [hunk 1]
```diff
@@ -138,3 +138,31 @@ def send_assigned_offer_email(
...
         return False
     return True
+
+
+def send_revoked_offer_email(template, learner_email, code):
+    """
+    Arguments:
+        *template*
+            The email template with placeholders that will receive the following tokens
+        *offer_assignment_id*
```
Comment (muhammad-ammar) [inline ecommerce/extensions/offer/utils.py]: `offer_assignment_id` is not in the function parameters list?

File: ecommerce/extensions/api/serializers.py [hunk 2]
```diff
@@ -1112,3 +1119,106 @@ def validate(self, data):
         data['voucher_usage_type'] = voucher_usage_type
         data['available_assignments'] = available_assignments
         return data
+
+
+class CouponCodeRevokeBulkSerializer(serializers.ListSerializer):  # pylint: disable=abstract-method
```
Comment (muhammad-ammar) [inline ecommerce/extensions/api/serializers.py:1169]: To me these customizations are complex to understand :smile: Any way to test this PR on sandbox?

File: ecommerce/extensions/api/v2/tests/views/test_enterprise.py [hunk 3]
```diff
@@ -823,3 +824,149 @@ def test_codes_assignment_email_failure(self, voucher_type, quantity, max_uses,
...
+            )
+
+        offer_assignment = OfferAssignment.objects.filter(user_email__in=emails).first()
+        with mock.patch('ecommerce.extensions.offer.utils.send_offer_update_email.delay') as mock_send_email:
+            response = self.get_response(
+                'POST',
+                '/api/v2/enterprise/coupons/{}/revoke/'.format(coupon_id),
+                [
+                    {'template': 'Test template', 'email': offer_assignment.user_email, 'code': offer_assignment.code},
+                    {'template': 'Test template', 'email': 'test3@example.com', 'code': 'RANDOMCODE'},
```
Comment (muhammad-ammar) [inline ecommerce/extensions/api/v2/tests/views/test_enterprise.py]: do we need to test a scenario where all bulk revokes completed successfully?

**Review by muhammad-ammar — COMMENTED** [review]
@brittneyexline Looks good to me. Any way to test these changes on sandbox?

File: ecommerce/extensions/api/serializers.py (hunk 2 above)
Comment (brittneyexline) [reply ecommerce/extensions/api/serializers.py:1169]: you are right, i need to add testing instructions :-)
i'll also add a comment about what these do - i took it from some previous work with the enterprise enrollment api where we needed to process multiple rows of request data independently, meaning failures wouldn't prevent other rows from being processed.

File: ecommerce/extensions/offer/utils.py [hunk 4]
```diff
@@ -99,3 +100,69 @@ def render_email_confirmation_if_required(request, offer, product):
...
+            Code for the user.
+        *enrollment_url*
+            URL for the user.
+        *code_usage_count*
+            Number of times the code can be redeemed.
+        *code_expiration_date*
+            Date till code is valid.
+
+    Returns:
+         True when successful or False in case of any exception
```
Comment (nedbat) [inline ecommerce/extensions/offer/utils.py]: This seems like an unusual choice for a function like this: why not let the exception raise up to the caller, and have them deal with it?

File: ecommerce/extensions/api/v2/views/enterprise.py [hunk 5]
```diff
@@ -242,10 +246,62 @@ def assign(self, request, pk):  # pylint: disable=unused-argument
...
+            )
+            status_keys = {
+                'id': assigned_offer.id,
+                'user_email': assigned_offer.user_email,
+                'code': assigned_offer.code,
+            }
+            if email_send_status:
+                status_keys.update({'status': 'email_dispatch_success'})
+            else:
+                status_keys.update({'status': 'email_dispatch_failed'})
```
Comment (nedbat) [inline ecommerce/extensions/api/v2/views/enterprise.py]: Wouldn't it be useful to include some of the reason why it failed?

File: ecommerce/extensions/offer/utils.py [hunk 6]
```diff
@@ -99,3 +100,69 @@ def render_email_confirmation_if_required(request, offer, product):
...
+        email_body = template.format(
+            code_usage_count=code_usage_count,
+            user_email=learner_email,
+            enrollment_url=enrollment_url,
+            code=code,
+            code_expiration_date=code_expiration_date
+        )
+        send_offer_assignment_email.delay(learner_email, offer_assignment_id, email_subject, email_body)
+    except Exception:  # pylint: disable=broad-except
+        return False
```
Comment (nedbat) [inline ecommerce/extensions/offer/utils.py]: We don't even log the exception? This completely discards all information about the failure.

**Review by nedbat — CHANGES_REQUESTED** [review]

File: ecommerce/extensions/api/v2/views/enterprise.py (hunk 5 above)
Comment (brittneyexline) [reply ecommerce/extensions/api/v2/views/enterprise.py]: this is actually part of another PR that my work is based on: https://github.com/edx/ecommerce/pull/2082
fyi @irfanuddinahmad - might not be applicable anymore for you though

File: ecommerce/extensions/offer/utils.py (hunk 4 above)
Comment (brittneyexline) [reply ecommerce/extensions/offer/utils.py]: also part of the PR mentioned above - but i added a similar function below so note taken :-)

File: ecommerce/extensions/offer/utils.py (hunk 6 above)
Comment (brittneyexline) [reply ecommerce/extensions/offer/utils.py]: this is also part of the PR mentioned above, but i added a similar function so note taken there!

File: ecommerce/extensions/api/serializers.py [hunk 7]
```diff
@@ -1158,3 +1164,114 @@ def _trigger_email_sending_task(self, template, assigned_offer, voucher_usage_ty
...
+        except Exception as exc:  # pylint: disable=broad-except
+            logger.exception('Encountered error when revoking code %s for user %s', code, email)
+            detail = unicode(exc)
+
+        validated_data['detail'] = detail
+        return validated_data
+
+    def validate(self, data):
+        """
+        Validate that the code is part of the Coupon the code and email provided have an active OfferAssignment.
```
Comment (irfanuddinahmad) [inline ecommerce/extensions/api/serializers.py]: nit: Validate that the code is part of the Coupon and the provided code and email have an active OfferAssignment.

File: ecommerce/extensions/offer/utils.py [hunk 8]
```diff
@@ -144,3 +144,26 @@ def send_assigned_offer_email(
...
+        *learner_email*
+            Email of the customer who will receive the code.
+        *code*
+            Code for the user.
+
+    Returns:
+         True when successful or False in case of any exception
+    """
+
+    email_subject = 'edX Course Assignment Revoked'
```
Comment (irfanuddinahmad) [inline ecommerce/extensions/offer/utils.py]: nit: Could be made a configuration setting.

File: ecommerce/extensions/offer/utils.py [hunk 9]
```diff
@@ -144,3 +144,26 @@ def send_assigned_offer_email(
...
+    Arguments:
+        *template*
+            The email template with placeholders that will receive the following tokens
+        *learner_email*
+            Email of the customer who will receive the code.
+        *code*
+            Code for the user.
+
+    Returns:
+         True when successful or False in case of any exception
```
Comment (irfanuddinahmad) [inline ecommerce/extensions/offer/utils.py]: Update docstring: The function does not return anything

**Review by irfanuddinahmad — APPROVED** [review]
LGTM! ... highlighted some small changes

**Pull request merged**